addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/MiHomeHandlerFactory.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.mihome.internal;
+
+import static org.openhab.binding.mihome.MiHomeBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Hashtable;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.openhab.binding.mihome.handler.MiHomeGatewayHandler;
+import org.openhab.binding.mihome.handler.MiHomeSubdevicesHandler;
+import org.openhab.binding.mihome.internal.discovery.MiHomeSubdeviceDiscoveryService;
+import org.osgi.framework.ServiceRegistration;
+
+import com.google.common.collect.Sets;
+
+/**
+ * The {@link MiHomeHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Svilen Valkanov - Initial contribution
+ */
+public class MiHomeHandlerFactory extends BaseThingHandlerFactory {
+
+    @SuppressWarnings("rawtypes")
```

Comment:
I might be mistaken but the class which is loaded as a dependency also has a method which is typed. We already use that method in org.openhab.binding.lutron.handler.IPBridgeHandler

I must agree that it is not described in OSGi 4.2 specification but first appears in [OSGi 4.3 specs](https://osgi.org/javadoc/r4v43/core/org/osgi/framework/BundleContext.html)

It allows for:

```java

    private Map<ThingUID, ServiceRegistration<DiscoveryService>> discoveryServiceRegs = new HashMap<>();

 private void registerDiscoveryService(MiHomeGatewayHandler gatewayHandler) {
        MiHomeSubdeviceDiscoveryService service = new MiHomeSubdeviceDiscoveryService(gatewayHandler);
        ServiceRegistration<DiscoveryService> serviceReg = bundleContext.registerService(DiscoveryService.class, service,
                new Hashtable<String, Object>());
        discoveryServiceRegs.put(gatewayHandler.getThing().getUID(), serviceReg);
    }
```

Looking at the manifest of org.eclipse.osgi-3.10.2.v20150203-1939.jar it states that it provides org.osgi.framework 1.8 => which is spec 6. 

The only thing I could find about version 4.2 is that Eclipse Smarthome runs on containers with osgi framework 4.2+ [[source](http://www.eclipse.org/smarthome/)]. However it seems that openHAB is running on a newer spec and restrictions for openhab are thus more relaxed than for Eclipse Smarthome with respect to the version of org.osgi.framework

@kaikreuzer is my assumption regarding the version of the osgi framework in openHAB vs Eclipse Smarthome correct?